CMakeLists.txt

```
@@ -123,6 +133,70 @@ if(USE_GPU)
     ADD_DEFINITIONS(-DUSE_GPU)
 endif(USE_GPU)
 
+if(USE_CUDA)
+    find_package(CUDA REQUIRED)
+    include_directories(${CUDA_INCLUDE_DIRS})
+    LIST(APPEND CMAKE_CUDA_FLAGS -g -Xcompiler=-fopenmp -Xcompiler=-fPIC -Xcompiler=-Wall -lineinfo)
+    CUDA_SELECT_NVCC_ARCH_FLAGS(CUDA_ARCH_FLAGS 7.0)
+
+    LIST(APPEND CMAKE_CUDA_FLAGS ${CUDA_ARCH_FLAGS})
+    if(CMAKE_BUILD_TYPE MATCHES Release)
+      LIST(APPEND CMAKE_CUDA_FLAGS -03)
+    endif()
+
+    message(STATUS "CMAKE_CUDA_FLAGS: ${CMAKE_CUDA_FLAGS}")
+    string(REPLACE ";" " " CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS}")
+    message(STATUS "CMAKE_CUDA_FLAGS: ${CMAKE_CUDA_FLAGS}")
+    set(CMAKE_CUDA_FLAGS_DEBUG -G)
+    set(CMAKE_CUDA_FLAGS_RELEASE -lineinfo)
```

We use our own `DEBUG` CMake option:
https://github.com/microsoft/LightGBM/blob/69a2691a599df7b259f2e79dc562895c8396c7d5/CMakeLists.txt#L15

Please change this piece of code to something like
```suggestion
    if(USE_DEBUG)
      SET(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -G")
    else()
      SET(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -O3 -lineinfo")
    endif()
    message(STATUS "CMAKE_CUDA_FLAGS: ${CMAKE_CUDA_FLAGS}")
```

Also, is `03` optimization level here? Then I guess it should be `O3` (letter `O`), not `03` (zero digit), refer to https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#options-for-altering-compiler-linker-behavior-optimize.

---

This line is needed

```
string(REPLACE ";" " " CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS}")
```

---

Sure, add any lines that are needed. I just showed the idea of our `DEBUG` flag.